File: src/main/java/com/back/global/rateLimit/Bucket4jRateLimiter.java
package com.back.global.rateLimit;

import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import io.github.bucket4j.Bandwidth;
import org.springframework.stereotype.Component;
import io.github.bucket4j.Bucket;
import io.github.bucket4j.Bandwidth;

import java.util.concurrent.TimeUnit;

@Component
public class Bucket4jRateLimiter {

    private final Cache<String, Bucket> cache;

    public Bucket4jRateLimiter() {
        this.cache = Caffeine.newBuilder()
                .expireAfterAccess(10, TimeUnit.MINUTES)
                .maximumSize(100_000)
                .build();
    }

    public boolean tryConsume(String key, RateLimitPolicy policy) {
        Bucket bucket = cache.get(key, k ->
                Bucket.builder()
                        .addLimit(Bandwidth.simple(
                                policy.getLimit(),
                                policy.getDuration()
                        ))
                        .build()
        );

        return bucket.tryConsume(1);
    }
}

File: src/main/java/com/back/global/rateLimit/Bucket4jRateLimiter.kt
package com.back.global.rateLimit

import com.github.benmanes.caffeine.cache.Cache
import com.github.benmanes.caffeine.cache.Caffeine
import io.github.bucket4j.Bandwidth
import io.github.bucket4j.Bucket
import org.springframework.stereotype.Component
import java.util.concurrent.TimeUnit
import java.util.function.Function

@Component
class Bucket4jRateLimiter {
    private val cache: Cache<String, Bucket> = Caffeine.newBuilder()
        .expireAfterAccess(10, TimeUnit.MINUTES)
        .maximumSize(100000)
        .build<String, Bucket>()


    fun tryConsume(key: String?, policy: RateLimitPolicy): Boolean =
        cache.get(key) {
            Bucket.builder()
                .addLimit(
                    Bandwidth.simple(
                        policy.limit.toLong(),
                        policy.duration
                    )
                )
                .build()
        }.tryConsume(1)

}


